### PR TITLE
[Concept Entry]interfaces go

### PR DESCRIPTION
**Pull request by Pavan-04-08**
### Description
continution to #1680 , closes https://github.com/Codecademy/docs/issues/1503

### Type of Change

- Adding a new entry

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [ ] I have performed a self-review of my own writing and code.
- [ ] I have checked my entry and corrected any misspellings.
- [ ] I have made corresponding changes to the documentation if needed.
- [ ] I have confirmed my changes are not being pushed from my forked `main` branch.
- [ ] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [ ] Under "Development" on the right, I have linked any issues that are relevant to this PR (write "Closes #<issue number> in the "Description" above).

### CODE REVIEW

File: content/go/concepts/interfaces/interfaces.md
Comment (SSwiniarski) [inline content/go/concepts/interfaces/interfaces.md:94]:
```suggestion
The advantage of using an interface is that any struct can be passed to the `newLife` function as long as it implements the `marriage` interface. and it is possible to  change the implementation of the `love` and `arranged` methods in the `john` and `max` structs without affecting the `newLife` function.
```

File: content/go/concepts/interfaces/interfaces.md
Comment (SSwiniarski) [inline content/go/concepts/interfaces/interfaces.md:96]:
```suggestion
It is also possible to make a new type and `newLife` can use it without changing the `newLife` function as long as the new type implements the `marriage` interface.
```

File: content/go/concepts/interfaces/interfaces.md
Comment (SSwiniarski) [inline content/go/concepts/interfaces/interfaces.md:108]:
```suggestion
The tradeoff of using an empty interface is the loss of type safety.  As there are no methods, it's possible to pass any type to a function that takes an empty interface as an argument.
```

File: content/go/concepts/interfaces/interfaces.md
Comment (SSwiniarski) [inline content/go/concepts/interfaces/interfaces.md:110]:
```suggestion
### Example

The `Println` function in the `fmt` package takes an empty interface as an argument.
```

File: content/go/concepts/interfaces/interfaces.md
Comment (SSwiniarski) [inline content/go/concepts/interfaces/interfaces.md:112]:
```suggestion
Below is the go source code for the `fmt.Println` function:
```

File: content/go/concepts/interfaces/interfaces.md
Comment (SSwiniarski) [inline content/go/concepts/interfaces/interfaces.md:122]:
```suggestion

> ##Note:** due to the empty interface, we can pass any type to the `Println` function.
```

File: content/go/concepts/interfaces/interfaces.md
Comment (SSwiniarski) [inline content/go/concepts/interfaces/interfaces.md:126]:
```suggestion
## Embedding an Interface
```

File: content/go/concepts/interfaces/interfaces.md
Comment (SSwiniarski) [inline content/go/concepts/interfaces/interfaces.md:124]: Move this line (124) to after the header at line 126.

File: content/go/concepts/interfaces/interfaces.md
Comment (SSwiniarski) [inline content/go/concepts/interfaces/interfaces.md:194]:
```suggestion
This example results in the following output:
```

File: content/go/concepts/interfaces/interfaces.md
Comment (SSwiniarski) [inline content/go/concepts/interfaces/interfaces.md:213]:
```suggestion
## Codebyte Example
```

File: content/go/concepts/interfaces/interfaces.md
Comment (SSwiniarski) [inline content/go/concepts/interfaces/interfaces.md:215]: There needs to be an actual interface example in the Codebyte, with an explanation of what it does.

**Review by SSwiniarski — CHANGES_REQUESTED** [review]
@Pavan-Nambi I did a first review and left several comments.

File: content/go/concepts/interfaces/interfaces.md
Comment (Pavan-Nambi) [reply content/go/concepts/interfaces/interfaces.md:218]: okay

File: content/go/concepts/interfaces/interfaces.md
Comment (SSwiniarski) [inline content/go/concepts/interfaces/interfaces.md:2]:
```suggestion
Title: 'Interfaces'
```

File: content/go/concepts/interfaces/interfaces.md
Comment (SSwiniarski) [inline content/go/concepts/interfaces/interfaces.md:90]:
```suggestion
- Structs `john` and `max` implement the `marriage` interface. Note that neither `john` nor `max` has any explicit declaration to implement the interface.
```

File: content/go/concepts/interfaces/interfaces.md
Comment (SSwiniarski) [inline content/go/concepts/interfaces/interfaces.md:92]:
```suggestion
- The function `newLife` takes a `marriage` interface as an argument.
```

File: content/go/concepts/interfaces/interfaces.md
Comment (SSwiniarski) [inline content/go/concepts/interfaces/interfaces.md:95]:
```suggestion
The advantage of using an interface is that any struct can be passed to the `newLife` function as long as it implements the `marriage` interface. It is possible to change the implementation of the `love` and `arranged` methods in the `john` and `max` structs without affecting the `newLife` function.
```

**Review by SSwiniarski — CHANGES_REQUESTED** [review]
@Pavan-Nambi, I have a few more comments to address.

**Comment by Pavan-Nambi** [comment]
@SSwiniarski  changed

**Review by SSwiniarski — APPROVED** [review]
@Pavan-Nambi, looks good for a second review.

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:126]:
```suggestion
> **Note:** Due to the empty interface, any type can be passed to the `Println()` function.
```

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:130]:
```suggestion
An interface in Go can be embedded in other interfaces or structures.
```

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:137]:
```suggestion
// Any type that needs to implement marriage should also implement all methods of dating.
```

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:172]: Fix any typos here and capitalize comments

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [reply content/go/concepts/interfaces/interfaces.md:221]: I think if we covered a simple interface in the previous examples and we want to include a code byte, we can display some extra functionality here.

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:219]: Can we combine this with the line above?

**Review by yangc95 — CHANGES_REQUESTED** [review]
Hey @Pavan-04-08 ! This is quite a lengthy entry, thanks for your efforts - I left some suggestions and comments. Please tag me if you need additional notes for requested changes.

File: content/go/concepts/interfaces/interfaces.md
Comment (Pavan-04-08) [reply content/go/concepts/interfaces/interfaces.md:167]: ok

File: content/go/concepts/interfaces/interfaces.md
Comment (Pavan-04-08) [reply content/go/concepts/interfaces/interfaces.md:216]: @yangc95  can you eloborate i didnt get this

Comment (Pavan-04-08) [reply content/go/concepts/interfaces/interfaces.md:216]: extra functionality hmm i didnt covered about few things like interface conversion,applied interfaces n stuff but if i just add code reg tht in codebyte without explaination it wont make sense...

what extra functionality are you referring to?

if you meant things like interface conversion n applied interfaces or anything more depth in interfaces i will make another PR after this as enhancement to interfaces and add things there with more explaination.

File: content/go/concepts/interfaces/interfaces.md
Comment (Pavan-04-08) [reply content/go/concepts/interfaces/interfaces.md:86]: isn't structure and struct basically same? even in official docs they refered it as `struct` https://go.dev/tour/moretypes/2

File: content/go/concepts/interfaces/interfaces.md
Comment (Pavan-04-08) [reply content/go/concepts/interfaces/interfaces.md:213]: yeah will change

File: content/go/concepts/interfaces/interfaces.md
Comment (Pavan-04-08) [reply content/go/concepts/interfaces/interfaces.md:25]: yeah will change

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [reply content/go/concepts/interfaces/interfaces.md:85]: Yeah they're definitively the same, I do suggest that if you want to mention structure as a keyword, you place them in backticks. So either use structure or `struct`. This is mainly a formatting preference for Docs

**Comment by SSwiniarski** [comment]
@yangc95 did @Pavan-04-08 address all your issues?

**Comment by Pavan-04-08** [comment]
> @yangc95 did @Pavan-04-08 address all your issues?

i was waiting for reply for this https://github.com/Codecademy/docs/pull/1708#discussion_r1146332652 
to update all at once..

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:15]:
```suggestion
An interface is composed of set of method signatures. These method signatures define the input and return values of which a data type or structure can conform to. In order to implement an interface, the type or structure must satisfy all the defined method signatures.
```

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:17]: Can we create an h2 here that is along the lines of "Creating an Interface" and nest the relevant code and text underneath

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [reply content/go/concepts/interfaces/interfaces.md:215]: Any enhancements would be great. As a base though, I think it would be good to set this concept up by breaking down how to create, and then implement interfaces. Within the create section, we can talk about syntax and give an example.

Then, in the implementation section, we can talk about how to implement an interface. Here, we can include the examples you've written so far with the addition of some meaningful explanation as to how the interface is being implemented. I think this is a better explanation.

**Comment by Pavan-04-08** [comment]
@yangc95  can you do a quick review for this?

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:15]:
```suggestion
##Creating an Interface
```

Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:15]: Can we move this? It should follow the introductory paragraph. Also, no need to use HTML formatting, we'll stick to markdown with the hashtags or pound symbol `##` (what I meant by h2)

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:17]: Here, I do mean structure as in [data structure](https://www.codecademy.com/resources/docs/general/data-structures), not `struct`.

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:27]: If this example is part of the creating an interface section, we'll use h3's

```suggestion
### Example
```

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:25]:
```suggestion
A type implements an interface by implementing its methods, In `Go`, interfaces are implemented implicitly and enables polymorphism.
```

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:69]:
```suggestion
package main

import (
  "fmt"
)

type marriage interface {
  love()
  arranged()
}

type john struct{}

func (a *john) love() {
  fmt.Println("love with john")
}

func (a *john) arranged() {
  fmt.Println("arranged marriage with john")
}

type max struct{}

func (a *max) love() {
  fmt.Println("love with max")
}

func (a *max) arranged() {
  fmt.Println("arranged marriage with max")
}

func newLife(guy marriage) {
  guy.love()
  guy.arranged()
}

func main() {
  newLife(&john{})
  newLife(&max{})
}
```

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:83]:
```suggestion
- The `marriage` interface has two methods: `love` and `arranged`.
```

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:85]:
```suggestion
- As long as the `struct` has all the methods in the interface, it is said to implement the interface.
```

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:87]:
```suggestion
- It's possible to call `newLife()` with the `john` and `max` structures.
```

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:89]:
```suggestion
The advantage of using an interface is that any `struct` can be passed to the `newLife()` function as long as it implements the `marriage` interface. It is possible to change the implementation of the `love` and `arranged` methods in the `john` and `max` structures without affecting the `newLife()` function.
```

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:189]:
```suggestion
```

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:184]:
```suggestion
    // It's possible to disquarely access the methods of embedded interface if the embedded interface is anonymous or unnamed.
```

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:176]:
```suggestion
    // p.love()
    // As this has named marriage interface m calling disquarely with .m would raise compiler error
    /*-----------------------ERROR-----*---------------------------*/
    // ./prog.go:39:4: p.love undefined (type parents has no field or method love)
```

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:164]:
```suggestion
    // Unnamed marriage interface
```

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:207]:
```suggestion
love with john
compatible with john
arranged marriage with john

I saw that you were perfect, and so I loved you. Then I saw that you were not perfect and I loved you even more

love with john
compatible with john
arranged marriage with john

Program exited.
```

**Review by yangc95 — COMMENTED** [review]
@Pavan-04-08 I left a few more comments. Thanks for your critical thinking and consistency with improving upon your entries ☺️

**Comment by Pavan-04-08** [comment]
is there anyway to know whats causing test to fail? ci tests are not rly giving any clue

File: content/go/concepts/interfaces/interfaces.md
Comment (yangc95) [inline content/go/concepts/interfaces/interfaces.md:25]: @Pavan-04-08 One more thing. Can we first go over an example of a simple implementation and move this part about polymorphism below? In the next iteration we can go into polymorphism a bit more

**Comment by yangc95** [comment]
> is there anyway to know whats causing test to fail? ci tests are not rly giving any clue

Looks like this is the issue:

Screenshot 2023-04-10 at 10 56 37 AM

File: content/go/concepts/interfaces/interfaces.md
Comment (Pavan-04-08) [inline content/go/concepts/interfaces/interfaces.md:81]: @yangc95  if you meant explaination to above example its given here.... heading is it just comes under creating an interface hm n yeah it makes more sense to move line abt polymorphism down will do tht

Comment (yangc95) [reply content/go/concepts/interfaces/interfaces.md:79]: @Pavan-04-08 This is more for pedagogical reasons. Creating an interface can be broken down into how to create an interface (with the interface keyword) and then how to implement an interface (wether that be done using a type or structure)

It would be good to show just one example first and keep only one `struct` (John or max) to better display this. Don't forget to adjust the example below to match!

If you disagree or have concerns, I'd love to discuss more

Comment (Pavan-04-08) [reply content/go/concepts/interfaces/interfaces.md:79]: @yangc95  ok... i got what you trying to say 
i would like to suggest something like https://www.codecademy.com/resources/docs/java/array-list
if we are going to keep things simpler doing like above like say array-list methods are not written in same page as rraylist definition 

likewise we can make interfaces too

on interfaces.md ---->interfaces explaintion with single struct n two structs
and then similar to arraylist methods at down on page
we can do for interfaces like say
empty interface
embedding interface with structs
interface conversion...etc

File: content/go/concepts/interfaces/interfaces.md
Comment (Pavan-04-08) [inline content/go/concepts/interfaces/interfaces.md:124]: @yangc95  one of the benefit for the approach i said at https://github.com/Codecademy/docs/pull/1708#discussion_r1162217190 
would be as you can see in the below example i just put embeding interface in anthr interface n struct in 1 single example we can divide this into two diff examples ....

File: content/go/concepts/interfaces/interfaces.md
Comment (Pavan-04-08) [reply content/go/concepts/interfaces/interfaces.md:79]: in this way we can be more details about things ig even if we dont add it would be like not overwhelming for a newbie who is lookin at docs here (in any case) as he go step by step instead of seeing a chunk abt interfaces at one place

Comment (yangc95) [reply content/go/concepts/interfaces/interfaces.md:79]: @Pavan-04-08 I think another issue here is that repeating the same examples (but with different names e.g., John and max) don't provide additional value or any new information here.

I can assist you with this change as well, if you give me a thumbs up 👍 

In addition, I noticed that you have many commits on this PR. If this is cumbersome, there's an option to add it to a batch, allowing you to push everything up in one commit. Hope this helps

**Comment by Pavan-04-08** [comment]
oo oops

**Comment by Pavan-04-08** [comment]
lol such a pain i messed up while rebasing 
will just change branch n make a new pr  tht shud be far easier

**Pull request closed without merging**
